### DISCUSSION TITLE
Dynamic Constraints for Many-to-Any Relationships in Directus

### DISCUSSION THREAD

**Discussion by bogdan-nita** [directus/directus#19502] [Feature Requests]
### Summary

Introduce dynamic constraint filters for many-to-any relationships in Directus to ensure content consistency within a group based on the type of the first item added.

### Basic Example

Basic Example
Imagine a Directus setup with a "blocks_group" collection that can reference any other collection, such as FAQ, Quotes, or Workshops.

Without the constraint filter:

A content editor begins populating a "blocks_group" by adding an FAQ item.
There's nothing stopping them from subsequently adding a Quote item to the same group, leading to mixed content types.

With the dynamic constraint filter:

An administrator sets up the "blocks_group" to have a many-to-any relationship with dynamic constraints based on the first item type added.
A content editor starts populating this "blocks_group" by adding an FAQ item.
Once the FAQ item is added, the system automatically applies a dynamic constraint, ensuring that only further FAQ items can be added to this group, preventing mix-ups.
If the editor decides to remove all items and begins with a Quote item instead, the constraint shifts to only allow Quote items for that group.

### Motivation

The need for this feature arises from the challenge of maintaining content consistency within Directus. In the absence of such dynamic constraints, there's a risk that content editors might inadvertently mix item types within the same group, leading to unpredictable content presentations on the frontend. A dynamic constraint would automatically guide content editors based on their initial item selection, ensuring uniformity within a group and avoiding potential content mix-ups.

### Detailed Design

Once a many-to-any relationship is set up, the system would monitor the first item type added to a group. Upon adding the first item, Directus would automatically apply a constraint to the group, restricting any subsequent additions to the same item type.

For instance, if an editor starts by adding an FAQ item to a "blocks_group", the system would dynamically set a constraint such that only FAQ items can be added to that group henceforth. If all items are removed, the constraint resets, allowing the editor to start fresh with a different item type.

### Must Have:
- System should detect the type of the first item added to a many-to-any group.
- Automatically apply a constraint based on the first item's type.
- Reset constraint when all items are removed from a group.
### Should Have:
- Clear feedback to content editors about the current constraint in effect.
- Option for administrators to override or modify constraints.
### Could Have:
- Option for administrators to set default constraints, even before the first item is added.
- Notifications/alerts for content editors when a constraint is applied or changed.
### Won't Have:
- Static constraints that don't change based on content.

### Drawbacks

While this feature enhances content consistency, it might introduce complexity for both administrators and content editors. There's a potential learning curve, and administrators might need to manage exceptions or scenarios where constraints need to be manually overridden. Furthermore, the introduction of this dynamic behavior might require thorough documentation to prevent confusion.

### Manual Oversight:
Rely on content editors being trained to maintain consistency, although this doesn't prevent accidental mix-ups.
### Separate Collections: 
Create separate collections for each type (e.g., blocks_faq, blocks_quotes). This method is more rigid and can lead to redundancy.
### Triggered Flow: 
Implement a flow that's triggered on the creation or editing of an item within the collection. This flow would check the content type of items already in the group and ensure that the new addition matches the existing content type. If not, the addition would be rejected or flagged for review. While this method enforces consistency, it might be reactive rather than proactive, potentially causing annoyance to content editors when their additions are denied post-hoc.
### Conditional Fields in Same Collection: 
Instead of separate collections or dynamic constraints, use a single collection but with separate fields for each content type (e.g., FAQ_field, Quotes_field). Based on the initial selection or the first item added, only the relevant field would be displayed to the content editor, while the others would be hidden. This ensures that only one type of content is added, based on the initially revealed field.

### Adoption Strategy

The feature can be introduced as an enhancement to the existing many-to-any relationship setup in Directus. Existing configurations remain unchanged until an administrator opts to use dynamic constraints. Providing thorough documentation, tutorials, and perhaps a toggle to enable/disable this feature can help in smooth adoption.

### Unresolved Questions

How would the system handle scenarios where manual overrides are necessary?
Are there performance implications, especially when dealing with large datasets and numerous item types?

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
